crates/m3/src/gadgets/u32/add.rs

```diff
@@ -134,6 +142,112 @@ impl U32Add {
 	}
 }
 
+/// A very simple trait used in Addition gadgets for unsigned integers of different bit lengths.
+pub trait UnsignedAddPrimitives {
+	type F: TowerField;
+	const BIT_LENGTH: usize;
+}
+
+impl UnsignedAddPrimitives for u32 {
+	type F = B32;
+
+	const BIT_LENGTH: usize = 32;
+}
+
+impl UnsignedAddPrimitives for u64 {
+	type F = B64;
+
+	const BIT_LENGTH: usize = 64;
+}
+
+/// Gadget for incrementing the values in a column by 1, generic over `u32` and `u64`
+#[derive(Debug)]
+pub struct Incr<UPrimitive: UnsignedAddPrimitives, const BIT_LENGTH: usize> {
+	cin: [Col<B1>; BIT_LENGTH],
+
+	/// Input column as bits.
+	pub input: [Col<B1>; BIT_LENGTH],
+	/// Output column as bits.
+	pub zout: [Col<B1>; BIT_LENGTH],
+	/// Output carry bit if there is any.
+	pub final_carry_out: Col<B1>,
+
+	_marker: PhantomData<UPrimitive>,
+}
+
+impl<UPrimitive: UnsignedAddPrimitives, const BIT_LENGTH: usize> Incr<UPrimitive, BIT_LENGTH> {
+	pub fn new(table: &mut TableBuilder, input: [Col<B1>; BIT_LENGTH]) -> Self {
+		assert_eq!(BIT_LENGTH, UPrimitive::BIT_LENGTH);
+		let cin = array::from_fn(|i| {
+			if i != 0 {
+				table.add_committed(format!("cout[{i}]"))
+			} else {
+				table.add_constant("cout[0]", [B1::ONE])
+			}
+		});
+		let zout = table.add_committed_multiple("zout");
+		let final_carry_out = table.add_committed("final_carry_out");
+		let cout: [_; BIT_LENGTH] = array::from_fn(|i| {
+			if i != BIT_LENGTH - 1 {
+				cin[i + 1]
+			} else {
+				final_carry_out
+			}
+		});
+
+		for (i, &xin) in input.iter().enumerate() {
+			table.assert_zero(format!("sum_{i}"), xin + cin[i] - zout[i]);
```

Comment:
Idk much about the conventions about feels like the indexing notation would be more consistent here.

```suggestion
			table.assert_zero(format!("sum[{i}]"), xin + cin[i] - zout[i]);
```